config.txt
#Thu Feb 08 16:14:40 CET 2018
cameraHori-droite.tempo=300
username=pi
jambes-bas.tempo=5000
brasLateralD-ecarter.tempo=100
poignetG-interieur.pin=B23
video_url=http\://192.168.1.15/cam_pic_new.php?time\=1515685109169&pDelay\=40000
onOffCartes.pin=5
mainD-fermer.tempo=500
cameraVert-haut.tempo=100
cameraEnabled=false
brasFrontalG-lever.tempo=3000
poignetD-exterieur.pin=B22
poignetD-interieur.tempo=200
busteIncli-avant.tempo=2500
diodesBuste.pin=B17
avantBrasG-baisser.pin=A13
busteRota-droite.pin=23
brasFrontalD-lever.pin=A28
avantBrasD-lever.tempo=4000
brasLateralD-ecarter.pin=A26
brasLateralG-ramener.pin=A15
brasFrontalD-baisser.tempo=1500
diodesBuste=B17
cameraVert-bas.tempo=100
cameraVert-haut.pin=25
mainG-ouvrir.pin=A11
poignetD-exterieur.tempo=200
ip_camera=192.168.1.15
lumiereYeux.pin=B15
busteIncli-arriere.pin=26
mainG-ouvrir.tempo=500
brasFrontalG-lever.pin=A18
tete-gauche.pin=4
avantBrasG-baisser.tempo=4000
paupieres-ouvrir.pin=B14
avantBrasG-lever.tempo=4000
cameraVert-bas.pin=22
mainG-fermer.pin=A12
mainG-fermer.tempo=500
brasLateralG-ramener.tempo=100
jambes-bas.pin=19
poignetD-interieur.pin=B21
cameraHori-gauche.pin=27
tete-gauche.tempo=500
busteIncli-arriere.tempo=6000
brasLateralG-ecarter.pin=A16
busteRota-gauche.tempo=1000
mainD-ouvrir.pin=A22
brasLateralG-ecarter.tempo=100
brasFrontalD-baisser.pin=A27
eclairage.pin=6
chassis-haut.pin=16
paupieres-ouvrir.tempo=800
chassis-bas.pin=13
paupieres-fermer.pin=B13
poignetG-interieur.tempo=200
poignetG-exterieur.pin=B24
ip_chariot=192.168.1.17
mainD-fermer.pin=A21
brasFrontalG-baisser.tempo=1750
avantBrasD-lever.pin=A24
tete-droite.pin=18
busteRota-gauche.pin=17
chassis-bas.tempo=10000
tete-droite.tempo=500
ip_corps=192.168.1.16
jambes-haut.tempo=7000
busteRota-droite.tempo=1000
cameraHori-gauche.tempo=300
brasFrontalG-baisser.pin=A17
password=Tengu70Valorun98
poignetG-exterieur.tempo=200
brasFrontalD-lever.tempo=2500
mainD-ouvrir.tempo=500
paupieres-fermer.tempo=800
avantBrasD-baisser.tempo=4000
busteIncli-avant.pin=21
brasLateralD-ramener.tempo=100
brasLateralD-ramener.pin=A25
avantBrasG-lever.pin=A14
jambes-haut.pin=20
cameraHori-droite.pin=24
chassis-haut.tempo=11000
avantBrasD-baisser.pin=A23


src/robot/Configuration.java
package robot;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.HashMap;
import java.util.Properties;

import com.jcraft.jsch.JSchException;

/**
 * Classe permettant de gérer le fichier de configuration.
 *
 * @author Ronan COLLIER
 *@version 1.0
 */

public class Configuration{
	public static Configuration instance;
	public static String videoUrl;
	public static boolean cameraEnabled;
	public static Raspberry rasPiCorps;
	public static Raspberry rasPiChar;
	public static Raspberry rasPiCam;
	public static File configFile;
	private Properties properties;
	public static HashMap<String, Relay> relaysMap;

	public Configuration() {
		if(instance==null)instance=this;
		configFile=new File(System.getProperty("user.dir")+"/config.txt");
		properties=new Properties();

		if(!configFile.exists()){ //si le fichier de config n'existe pas, on en creer un par défaut
		}
		try {
			//charge le fichier de propriétés
			properties.load(new FileInputStream(configFile));
		}
		catch (IOException e) {
			e.printStackTrace();
		}
		readConnectionsProperties();
		readRelaysProperties();
		readMiscProperties();
		try {
			properties.store(new FileOutputStream(configFile), null);
		} catch (IOException e) {
			e.printStackTrace();
		}
		uploadConfigFile();
	}
	private void readConnectionsProperties() {
		//cherche la clé correspondante, et lui donne une valeur par défaut si la clé n'est pas trouvée.
		String username=getProperty("username", "pi");
		String password=getProperty("password", "raspberry");
		String ipCorps=getProperty("ip_corps", "192.168.1.16");
		String ipChar=getProperty("ip_chariot", "192.168.1.17");
		String ipCam=getProperty("ip_camera", "192.168.1.15");
		videoUrl=getProperty("video_url");
		System.out.println(videoUrl);
		System.out.println(ipCorps);

		rasPiCorps=new Raspberry(ipCorps, username, password);
		rasPiChar=new Raspberry(ipChar, username, password);
		rasPiCam=new Raspberry(ipCam, username, password);
	}
	private void readRelaysProperties() {
		relaysMap = new HashMap<String, Relay>();
		getRelay("tete-gauche", "4", 500);
		getRelay("tete-droite", "18", 400);
		getRelay("busteRota-gauche", "17", 1000);
		getRelay("busteRota-droite", "23", 1000);
		getRelay("cameraHori-gauche", "27", 300);
		getRelay("cameraHori-droite", "24", 300);
		getRelay("cameraVert-bas", "22", 100);
		getRelay("cameraVert-haut", "25", 100);
		getRelay("chassis-bas", "13", 10000);
		getRelay("chassis-haut", "16", 11000);
		getRelay("jambes-bas", "19", 5000);
		getRelay("jambes-haut", "20", 7000);
		getRelay("busteIncli-arriere", "26", 6000);
		getRelay("busteIncli-avant", "21", 2500);

		getRelay("onOffCartes", "5");
		getRelay("eclairage", "6");
		getRelay("lumiereYeux", "B15");
		getRelay("diodesBuste", "B17");
		getRelay("paupieres-fermer", "B13", 600);
		getRelay("paupieres-ouvrir", "B14", 600);
		/*##### pins bras gauche #####*/
		getRelay("brasFrontalG-lever", "A18", 3000);
		getRelay("brasFrontalG-baisser", "A17",1750);
		getRelay("brasLateralG-ecarter", "A16", 100);
		getRelay("brasLateralG-ramener", "A15", 100);
		getRelay("avantBrasG-lever", "A14", 4000);
		getRelay("avantBrasG-baisser", "A13", 4000);
		getRelay("mainG-fermer", "A12", 500);
		getRelay("mainG-ouvrir", "A11", 500);
		getRelay("poignetG-exterieur", "B24", 200);
		getRelay("poignetG-interieur", "B23", 200);
		/*##### pins bras droit #####*/
		getRelay("brasFrontalD-lever", "A28", 2500);
		getRelay("brasFrontalD-baisser", "A27", 1500);
		getRelay("brasLateralD-ecarter", "A26", 100);
		getRelay("brasLateralD-ramener", "A25", 100);
		getRelay("avantBrasD-lever", "A24", 4000);
		getRelay("avantBrasD-baisser", "A23", 4000);
		getRelay("mainD-fermer", "A22", 500);
		getRelay("mainD-ouvrir", "A21", 500);
		getRelay("poignetD-exterieur", "B22", 200);
		getRelay("poignetD-interieur", "B21", 200);

	}
	private void readMiscProperties() {
		cameraEnabled=Boolean.parseBoolean(getProperty("cameraEnabled", "true"));
	}
	/**
	 * Méthode permettant de lire une propriété dans le fichier de configuration.
	 * @param key la clé identifiant la propriété recherchée.
	 * @return la valeur de la propriété.
	 */
	private String getProperty(String key){
		String value = properties.getProperty(key);
		return value;
	}
	/**
	 * Méthode permettant de lire une propriété dans le fichier de configuration, et lui attribut une valeur si la propriété n'existe pas.
	 * @param key la clé identifiant la propriété recherchée.
	 * @param defaultVal la valeur par défaut assigné à la propriété si cette dernière n'existe pas.
	 * @return la valeur de la propriété.
	 */
	private String getProperty(String key, String defaultVal){
		String value = properties.getProperty(key, defaultVal);
		if(!properties.containsKey(key)) {
			properties.setProperty(key, defaultVal);
			System.out.println(">>>Config file edited.");
		}
		return value;
	}
	/**
	 * Méthode permettant de rechercher un relai dans le fichier de configuration, et l'ajoute dans la liste des relais.
	 * @param relayName le nom du relai désiré.
	 * @param defaultPin le pin par défaut du relai si ce dernier n'est pas présent dans le fichier de configuration.
	 */
	private void getRelay(String relayName, String defaultPin ){
		Relay relay=new Relay(relayName, getProperty(relayName+".pin",defaultPin));
		relaysMap.put(relayName, relay);
	}
	/**
	 * Méthode permettant de rechercher un relai temporisé dans le fichier de configuration, et l'ajoute dans la liste des relais.
	 * @param relayName le nom du relai désiré.
	 * @param defaultPin le pin par défaut du relai si le pin n'est pas présent dans le fichier de configuration.
	 * @param defaultTempo la temporisation par défaut du relai si la tempo n'est pas présente dans le fichier de configuration.
	 */
	private void getRelay(String relayName, String defaultPin, int defaultTempo ){
		TemporizedRelay relay=new TemporizedRelay(relayName, getProperty(relayName+".pin",defaultPin),
				Integer.parseInt(getProperty(relayName+".tempo",defaultTempo+"")));
		relaysMap.put(relayName, relay);
	}
	/**
	 * Méthode permettant d'obtenir la liste des relais sous forme d'une hasmap de la forme |nom_du_relai, relai|.
	 * @return la liste des relais sous la forme d'une hashmap.
	 */
	public HashMap<String, Relay> getRelaysMap(){
		return relaysMap;
	}
	/**
	 * Méthode permettant de télécharer le fichier de configuration sur le raspberry du corps.
	 */
	public void uploadConfigFile() {
		String str="";
		if(rasPiCorps.checkConnection()) {
			for (String key : relaysMap.keySet()) {
				Relay value = relaysMap.get(key);
				str+=key+".pin="+value.getPin()+"\n";
				if(TemporizedRelay.class.isInstance(value))str+=key+".tempo="+((TemporizedRelay)value).getTempo()+"\n";
			}

			System.out.println(str);
			rasPiCorps.sendCommand("echo \'"+str+"\' > /home/pi/Scripts/config.txt");
		}


	}
	/**
	 * Méthode permettant de mettre à jour l'état d'une connection. Si la connexion est établie, on l'éteint, sinon on l'allume.
	 * @param connection le nom de la connection dont on veut mettre à jour l'état.
	 */
	public void updateConnections(String connection) {
		if (connection=="Connexion raspberry corps") {
			if(rasPiCorps.checkConnection()==true) {
				rasPiCorps.disconnect();
			}
			else{
				try {
					rasPiCorps.connect();
					uploadConfigFile();
				}
				catch(JSchException | IOException ex) {
					System.err.println("Error: connection failed");
					ex.printStackTrace();
				}
			}
		}

		else if (connection=="Connexion raspberry chariot") {
			if(rasPiChar.checkConnection()==true) {
				rasPiChar.disconnect();
			}
			else{
				try {
					rasPiChar.connect();
				}
				catch(JSchException | IOException ex) {
					System.err.println("Error: connection failed");
					ex.printStackTrace();
				}
			}
		}
		else if (connection=="Connexion raspberry caméra") {
			if(rasPiCam.checkConnection()==true) {
				rasPiCam.disconnect();
			}
			else{
				try {
					rasPiCam.connect();
				}
				catch(JSchException | IOException ex) {
					System.err.println("Error: connection failed");
					ex.printStackTrace();
				}
			}
		}
	}

}